File: tests/cpp/test_shrinkwrap.cpp
//  ██████   ██████ ██████████  █████████  █████   █████ █████    ███████
// ░░██████ ██████ ░░███░░░░░█ ███░░░░░███░░███   ░░███ ░░███   ███░░░░░███      ███         ███
//  ░███░█████░███  ░███  █ ░ ░███    ░░░  ░███    ░███  ░███  ███     ░░███    ░███        ░███
//  ░███░░███ ░███  ░██████   ░░█████████  ░███████████  ░███ ░███      ░███ ███████████ ███████████
//  ░███ ░░░  ░███  ░███░░█    ░░░░░░░░███ ░███░░░░░███  ░███ ░███      ░███░░░░░███░░░ ░░░░░███░░░
//  ░███      ░███  ░███ ░   █ ███    ░███ ░███    ░███  ░███ ░░███     ███     ░███        ░███
//  █████     █████ ██████████░░█████████  █████   █████ █████ ░░░███████░      ░░░         ░░░
// ░░░░░     ░░░░░ ░░░░░░░░░░  ░░░░░░░░░  ░░░░░   ░░░░░ ░░░░░    ░░░░░░░
//
//
//  License:         MIT License
//                   meshio++ default license: LICENSE
//
//  Main authors:    Vicente Mataix Ferrandiz
//
//
// Tests for shrinkwrap. See operations/shrinkwrap.hpp for the contract; the
// oracles here are stated in the plan that shipped it: projection onto a
// sphere measures zero distance afterwards, a planar target with an offset
// puts every point exactly at that height, the accelerator is unobservable,
// and a crease offsets along the bisector (the one deliberate divergence from
// upstream, sabotage-verified by switching to the face normal).

// System includes
#include <cmath>
#include <cstdint>
#include <vector>

// External includes
#include <gtest/gtest.h>

// Project includes
#include "meshioplusplus/operations/sdf.hpp"
#include "meshioplusplus/operations/shrinkwrap.hpp"
#include "mesh_fixtures.hpp"

using namespace meshioplusplus;

namespace {

using mt::icosphere;
using mt::plane_grid;

/// A point cloud (vertex cells) at the given positions.
Mesh cloud(const std::vector<std::vector<double>>& rPts) {
    std::vector<std::vector<std::int64_t>> v;
    for (std::size_t i = 0; i < rPts.size(); ++i)
        v.push_back({static_cast<std::int64_t>(i)});
    return mt::make_mesh(rPts, "vertex", v);
}

/// Two unit squares meeting at the x axis at 90 degrees (a "book"): one in the
/// z = 0 plane (y >= 0), one in the y = 0 plane (z >= 0), both wound so the
/// normals point into the +y/+z quadrant. The spine is the x axis.
Mesh book() {
    std::vector<std::vector<double>> p = {{0, 0, 0}, {1, 0, 0}, {1, 1, 0},
                                          {0, 1, 0}, {1, 0, 1}, {0, 0, 1}};
    std::vector<std::vector<std::int64_t>> f = {{0, 1, 2}, {0, 2, 3}, {0, 4, 1}, {0, 5, 4}};
    return mt::make_mesh(p, "triangle", f);
}

std::vector<double> coords(const Mesh& rMesh) {
    const NDArray& p = rMesh.Points();
    std::vector<double> out(p.Size());
    for (std::size_t i = 0; i < p.Size(); ++i)
        out[i] = detail::read_double(p, i);
    return out;
}

}  // namespace

TEST(Shrinkwrap, ProjectsOntoASphereSoTheDistanceIsZeroAfterwards) {
    const Mesh target = icosphere(3, 1.0);
    const Mesh source = icosphere(2, 1.3);
    const ShrinkwrapResult r = shrinkwrap(source, target);
    EXPECT_EQ(r.mNumProjected, static_cast<std::int64_t>(source.NumPoints()));
    EXPECT_EQ(r.mNumMissed, 0);
    EXPECT_EQ(r.mNumSkipped, 0);
    EXPECT_NEAR(r.mMaxDisplacement, 0.3, 0.02);
    EXPECT_TRUE(r.mQuality.mWatertight);
    const NDArray d = sample_distance(target, r.mMesh.Points());
    for (std::size_t i = 0; i < d.Size(); ++i)
        EXPECT_NEAR(detail::read_double(d, i), 0.0, 1e-12);
    // Connectivity and cell count untouched: a pure coordinate move.
    EXPECT_EQ(r.mMesh.NumCellBlocks(), source.NumCellBlocks());
    EXPECT_EQ(r.mMesh.Cells(0).NumCells(), source.Cells(0).NumCells());
}

TEST(Shrinkwrap, OffsetOnAPlanePutsEveryPointAtThatHeight) {
    const Mesh target = plane_grid(4);  // z = 0, wound +z
    std::vector<std::vector<double>> pts;
    for (int i = 0; i < 12; ++i)
        pts.push_back({0.7 + 0.2 * i, 1.3 + 0.1 * i, (i % 3 == 0 ? -1.0 : 1.0) * (0.1 + 0.05 * i)});
    const Mesh source = cloud(pts);
    ShrinkwrapOptions o;
    o.mOffset = 0.25;
    const ShrinkwrapResult r = shrinkwrap(source, target, o);
    const std::vector<double> c = coords(r.mMesh);
    double max_disp = 0.0;
    for (std::size_t i = 0; i < pts.size(); ++i) {
        EXPECT_DOUBLE_EQ(c[i * 3 + 0], pts[i][0]);
        EXPECT_DOUBLE_EQ(c[i * 3 + 1], pts[i][1]);
        EXPECT_NEAR(c[i * 3 + 2], 0.25, 1e-15);
        max_disp = std::max(max_disp, std::fabs(pts[i][2] - 0.25));
    }
    EXPECT_NEAR(r.mMaxDisplacement, max_disp, 1e-15);
}

TEST(Shrinkwrap, MaxDistanceLeavesFarPointsAloneAndCountsThem) {
    const Mesh target = plane_grid(4);
    const Mesh source = cloud({{1, 1, 0.05}, {2, 2, 0.5}, {1.5, 2.5, -0.02}, {3, 1, -2.0}});
    ShrinkwrapOptions o;
    o.mMaxDistance = 0.1;
    o.mRecordDistance = true;
    o.mRecordClosestCell = true;
    const ShrinkwrapResult r = shrinkwrap(source, target, o);
    EXPECT_EQ(r.mNumProjected, 2);
    EXPECT_EQ(r.mNumMissed, 2);
    const std::vector<double> c = coords(r.mMesh);
    EXPECT_DOUBLE_EQ(c[2], 0.0);
    EXPECT_DOUBLE_EQ(c[5], 0.5);  // untouched
    EXPECT_DOUBLE_EQ(c[8], 0.0);
    EXPECT_DOUBLE_EQ(c[11], -2.0);  // untouched
    // The distance is recorded for every queried point, hit or miss.
    const double* d = r.mMesh.PointData(kShrinkwrapDistanceName).As<double>();
    EXPECT_NEAR(d[1], 0.5, 1e-15);
    EXPECT_NEAR(d[3], 2.0, 1e-15);
    const std::int64_t* cc = r.mMesh.PointData(kShrinkwrapClosestCellName).As<std::int64_t>();
    EXPECT_GE(cc[0], 0);
    EXPECT_GE(cc[3], 0);
}

TEST(Shrinkwrap, WeightsBlendOrSelect) {
    const Mesh target = plane_grid(4);
    Mesh source = cloud({{1, 1, 1.0}, {2, 2, 1.0}, {3, 3, 1.0}});
    // Float weights blend, bit-exactly x + w*(p - x).
    source.AddPointData("w", mt::data_array({0.5, 0.0, 1.0}));
    ShrinkwrapOptions o;
    o.mWeights = "w";
    const ShrinkwrapResult r = shrinkwrap(source, target, o);
    const std::vector<double> c = coords(r.mMesh);
    EXPECT_DOUBLE_EQ(c[2], 1.0 + 0.5 * (0.0 - 1.0));
    EXPECT_DOUBLE_EQ(c[5], 1.0);  // weight 0: never queried
    EXPECT_DOUBLE_EQ(c[8], 0.0);
    EXPECT_EQ(r.mNumSkipped, 1);
    EXPECT_EQ(r.mNumProjected, 2);
    // Integer weights select.
    Mesh sel = cloud({{1, 1, 1.0}, {2, 2, 1.0}});
    sel.AddPointData("pick", mt::int_data_array({0, 1}));
    o.mWeights = "pick";
    const ShrinkwrapResult r2 = shrinkwrap(sel, target, o);
    const std::vector<double> c2 = coords(r2.mMesh);
    EXPECT_DOUBLE_EQ(c2[2], 1.0);
    EXPECT_DOUBLE_EQ(c2[5], 0.0);
    EXPECT_EQ(r2.mNumSkipped, 1);
    // A wrong name is refused by name.
    o.mWeights = "nope";
    EXPECT_THROW(shrinkwrap(sel, target, o), std::invalid_argument);
}

TEST(Shrinkwrap, TheBucketSizeDoesNotChangeTheAnswer) {
    const Mesh target = icosphere(2, 1.0);
    const Mesh source = icosphere(1, 1.4);
    ShrinkwrapOptions o;
    o.mOffset = 0.05;
    const std::vector<double> ref = coords(shrinkwrap(source, target, o).mMesh);
    for (const double cell : {0.05, 0.3, 5.0}) {
        o.mGridCellSize = cell;
        const ShrinkwrapResult r = shrinkwrap(source, target, o);
        const std::vector<double> c = coords(r.mMesh);
        ASSERT_EQ(c.size(), ref.size());
        for (std::size_t i = 0; i < c.size(); ++i)
            EXPECT_EQ(c[i], ref[i]) << "bucket " << cell << " entry " << i;
    }
}

// The one deliberate divergence from upstream. A point on the CONVEX side of
// the spine, on its bisector, projects onto the spine (an EDGE hit, two
// equidistant faces -- on the concave side the nearest feature would be a
// face), and
// the offset must go along the bisector -- the feature pseudonormal -- not
// along whichever face won the tie-break. Sabotaging the kernel to return the
// face normal moves the point along one face and fails this.
TEST(Shrinkwrap, ACreaseOffsetsAlongTheBisector) {
    const Mesh target = book();
    const double s = 0.5 / std::sqrt(2.0);
    const Mesh source = cloud({{0.5, -s, -s}});
    ShrinkwrapOptions o;
    o.mOffset = 0.1;
    const ShrinkwrapResult r = shrinkwrap(source, target, o);
    const std::vector<double> c = coords(r.mMesh);
    const double b = 0.1 / std::sqrt(2.0);
    EXPECT_NEAR(c[0], 0.5, 1e-15);
    EXPECT_NEAR(c[1], b, 1e-15);
    EXPECT_NEAR(c[2], b, 1e-15);
    // Under area weighting the same answer: the edge normal is the plain sum
    // of the two unit face normals under either weighting.
    o.mNormalWeight = SdfPseudonormalWeight::Area;
    const std::vector<double> c2 = coords(shrinkwrap(source, target, o).mMesh);
    EXPECT_NEAR(c2[1], b, 1e-15);
    EXPECT_NEAR(c2[2], b, 1e-15);
}

TEST(Shrinkwrap, AVolumeSourceMovesEveryPointIncludingInteriorOnes) {
    const Mesh target = icosphere(3, 1.0);
    // A tetra whose corners are all inside the unit sphere.
    const Mesh source = mt::tet_mesh();
    const ShrinkwrapResult r = shrinkwrap(source, target);
    EXPECT_EQ(r.mNumProjected, static_cast<std::int64_t>(source.NumPoints()));
    const NDArray d = sample_distance(target, r.mMesh.Points());
    for (std::size_t i = 0; i < d.Size(); ++i)
        EXPECT_NEAR(detail::read_double(d, i), 0.0, 1e-12);
    EXPECT_EQ(std::string(r.mMesh.Cells(0).Type()), "tetra");
}

TEST(Shrinkwrap, DataRegionsAndDtypeSurvive) {
    const Mesh target = plane_grid(3);
    Mesh source = mt::data_mesh();
    const ShrinkwrapResult r = shrinkwrap(source, target);
    EXPECT_EQ(r.mMesh.Points().Dtype(), source.Points().Dtype());
    EXPECT_EQ(r.mMesh.PointDataNames(), source.PointDataNames());
    EXPECT_EQ(r.mMesh.CellDataNames(), source.CellDataNames());
    EXPECT_EQ(r.mMesh.NumRegions(), source.NumRegions());
}

TEST(Shrinkwrap, RefusesAVolumeTargetByName) {
    const Mesh source = mt::tri_mesh();
    try {
        shrinkwrap(source, mt::tet_mesh());
        FAIL() << "expected a throw";
    } catch (const std::invalid_argument& e) {
        const std::string what = e.what();
        EXPECT_NE(what.find("meshio++: shrinkwrap: target:"), std::string::npos) << what;
        EXPECT_NE(what.find("extract_surface"), std::string::npos) << what;
    }
}

File: tests/python/test_shrinkwrap.py
"""Tests for ``shrinkwrap``. Fixtures mirror ``tests/cpp/test_shrinkwrap.cpp``;
the numpy twin is pinned bit-exact against the core wherever the module
docstring promises it, and refuses by name where it does not."""

import math

import numpy as np
import pytest

import meshioplusplus as mio
from meshioplusplus import sample_distance, shrinkwrap
from meshioplusplus._shrinkwrap import CLOSEST_CELL_NAME, DISTANCE_NAME, _shrinkwrap_py

from .test_curvature import icosphere, plane_grid

try:
    from meshioplusplus import _core
except ImportError:  # pragma: no cover - a pure-Python build
    _core = None

needs_core = pytest.mark.skipif(_core is None, reason="parity needs the compiled core")


def cloud(pts):
    pts = np.asarray(pts, dtype=np.float64)
    return mio.Mesh(pts, [("vertex", np.arange(len(pts)).reshape(-1, 1))])


def book():
    pts = np.array(
        [[0, 0, 0], [1, 0, 0], [1, 1, 0], [0, 1, 0], [1, 0, 1], [0, 0, 1]], float
    )
    f = np.array([[0, 1, 2], [0, 2, 3], [0, 4, 1], [0, 5, 4]], dtype=np.int64)
    return mio.Mesh(pts, [("triangle", f)])


def test_projects_onto_a_sphere():
    target = icosphere(3)
    source = icosphere(2, radius=1.3)
    out, rep = shrinkwrap(source, target, return_report=True)
    assert rep["num_projected"] == len(source.points)
    assert rep["num_missed"] == 0 and rep["num_skipped"] == 0
    assert rep["max_displacement"] == pytest.approx(0.3, abs=0.02)
    assert rep["quality"]["watertight"]
    d = sample_distance(target, out.points)
    np.testing.assert_allclose(d, 0.0, atol=1e-12)
    assert [cb.type for cb in out.cells] == ["triangle"]
    np.testing.assert_array_equal(out.cells[0].data, source.cells[0].data)


def test_offset_on_a_plane_puts_every_point_at_that_height():
    target = plane_grid(4)
    pts = [
        (0.7 + 0.2 * i, 1.3 + 0.1 * i, (-1 if i % 3 == 0 else 1) * (0.1 + 0.05 * i))
        for i in range(12)
    ]
    out = shrinkwrap(cloud(pts), target, offset=0.25)
    np.testing.assert_allclose(out.points[:, 2], 0.25, atol=1e-15)
    np.testing.assert_array_equal(out.points[:, :2], np.asarray(pts)[:, :2])


def test_max_distance_and_recorded_arrays():
    target = plane_grid(4)
    source = cloud([[1, 1, 0.05], [2, 2, 0.5], [1.5, 2.5, -0.02], [3, 1, -2.0]])
    out, rep = shrinkwrap(
        source,
        target,
        max_distance=0.1,
        record_distance=True,
        record_closest_cell=True,
        return_report=True,
    )
    assert rep["num_projected"] == 2 and rep["num_missed"] == 2
    np.testing.assert_array_equal(out.points[:, 2], [0.0, 0.5, 0.0, -2.0])
    d = out.point_data[DISTANCE_NAME]
    assert d[1] == pytest.approx(0.5) and d[3] == pytest.approx(2.0)
    assert (out.point_data[CLOSEST_CELL_NAME] >= 0).all()


def test_weights_blend_or_select_by_name_or_array():
    target = plane_grid(4)
    source = cloud([[1, 1, 1.0], [2, 2, 1.0], [3, 3, 1.0]])
    source.point_data["w"] = np.array([0.5, 0.0, 1.0])
    out, rep = shrinkwrap(source, target, weights="w", return_report=True)
    np.testing.assert_array_equal(out.points[:, 2], [0.5, 1.0, 0.0])
    assert rep["num_skipped"] == 1
    # An array is attached temporarily and never leaks into the output.
    out2 = shrinkwrap(source, target, weights=np.array([0, 1, 1]))
    np.testing.assert_array_equal(out2.points[:, 2], [1.0, 0.0, 0.0])
    assert "shrinkwrap:weights" not in out2.point_data
    with pytest.raises(ValueError, match="nope"):
        shrinkwrap(source, target, weights="nope")


def test_a_crease_offsets_along_the_bisector():
    """The deliberate divergence from upstream: an edge hit offsets along the
    feature pseudonormal (the bisector), not along one face's normal."""
    s = 0.5 / math.sqrt(2.0)
    out = shrinkwrap(cloud([[0.5, -s, -s]]), book(), offset=0.1)
    b = 0.1 / math.sqrt(2.0)
    np.testing.assert_allclose(out.points[0], [0.5, b, b], atol=1e-15)


def test_a_volume_source_moves_interior_points_too():
    target = icosphere(3)
    tet = mio.Mesh(
        np.array([[0, 0, 0], [0.5, 0, 0], [0, 0.5, 0], [0, 0, 0.5]], float),
        [("tetra", np.array([[0, 1, 2, 3]]))],
    )
    out, rep = shrinkwrap(tet, target, return_report=True)
    assert rep["num_projected"] == 4
    np.testing.assert_allclose(sample_distance(target, out.points), 0.0, atol=1e-12)
    assert out.cells[0].type == "tetra"


def test_data_and_sets_survive_and_the_target_is_checked():
    target = plane_grid(3)
    m = icosphere(0)
    m.point_data["T"] = np.arange(12, dtype=float)
    m.point_sets["s"] = np.array([1, 2])
    out = shrinkwrap(m, target)
    np.testing.assert_array_equal(out.point_data["T"], m.point_data["T"])
    assert sorted(out.point_sets["s"]) == [1, 2]
    assert out.points.dtype == m.points.dtype
    tet = mio.Mesh(np.eye(4, 3), [("tetra", np.array([[0, 1, 2, 3]]))])
    with pytest.raises(ValueError, match="shrinkwrap: target"):
        shrinkwrap(m, tet)


def test_numpy_twin_refuses_an_angle_weighted_offset():
    with pytest.raises(NotImplementedError, match="area"):
        _shrinkwrap_py(
            cloud([[0, 0, 1]]), plane_grid(2), 0.1, 0.0, "", "", "angle", False, False
        )
    # Offset 0 under "angle", and any offset under "area", are fine.
    _shrinkwrap_py(
        cloud([[0.5, 0.5, 1]]), plane_grid(2), 0.0, 0.0, "", "", "angle", False, False
    )
    _shrinkwrap_py(
        cloud([[0.5, 0.5, 1]]), plane_grid(2), 0.1, 0.0, "", "", "area", False, False
    )


@needs_core
@pytest.mark.parametrize(
    "offset, weight", [(0.0, "angle"), (0.0, "area"), (0.07, "area"), (-0.2, "area")]
)
def test_cpp_matches_python(offset, weight):
    """Exact equality: the projection is + - * / and one sqrt over the same
    totally ordered search the SDF twin already pins."""
    target = icosphere(2)
    source = icosphere(1, radius=1.35)
    source.point_data["w"] = np.linspace(0.2, 1.0, len(source.points))
    for weights in ("", "w"):
        cpp, cpp_rep = shrinkwrap(
            source,
            target,
            offset=offset,
            max_distance=0.5,
            weights=weights or None,
            normal_weight=weight,
            record_distance=True,
            record_closest_cell=True,
            return_report=True,
        )
        py, py_rep = _shrinkwrap_py(
            source, target, offset, 0.5, weights, "", weight, True, True
        )
        np.testing.assert_array_equal(cpp.points, py.points)
        np.testing.assert_array_equal(
            cpp.point_data[DISTANCE_NAME], py.point_data[DISTANCE_NAME]
        )
        np.testing.assert_array_equal(
            cpp.point_data[CLOSEST_CELL_NAME], py.point_data[CLOSEST_CELL_NAME]
        )
        for k in (
            "num_projected",
            "num_missed",
            "num_skipped",
            "max_displacement",
            "quality",
        ):
            assert cpp_rep[k] == py_rep[k], k


@needs_core
def test_cpp_matches_python_on_the_crease():
    src = cloud([[0.5, -0.3, -0.3], [0.2, -0.1, 0.4], [0.9, 0.4, -0.05]])
    cpp = shrinkwrap(src, book(), offset=0.1, normal_weight="area")
    py, _ = _shrinkwrap_py(src, book(), 0.1, 0.0, "", "", "area", False, False)
    np.testing.assert_array_equal(cpp.points, py.points)
